Mypy annotation zerver.lib.event_queue and zerver.lib.queue

cpdean opened the pull request:
I ran into some problems with send_event taking either a list of ints or a list of dicts, so i figured i would change the caller's code.

Had problems with events either being a Dict with str keys and one caller who put an int key in there.

Additionally, needed to create some placeholder callbacks where mypy claimed it couldn't infer the type of lambdas.

smarx commented:
*Automated message from Dropbox CLA bot*

@cpdean, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
Cool, thanks for doing this @cpdean!  I merged versions of the first 3 commits (mostly just clarifying commit messages, plus a bit of cleanup to a27e50c70d5d6d675cdaafbd438cf69258ff211d to make it pass mypy checks and remove annotations not directly related to the change).

Can you rebase this branch to squash the fixes to bugs introduced in this branch into the original commit that introduced them?  Then it'll be a lot easier to incrementally merge the remaining pieces of this.  
check out the Zulip coding style guidelines here for details on our commit style: https://zulip.readthedocs.org/en/latest/code-style.html#version-control

Also, I'm curious about the need for `nop_callback` -- is that working around a mypy bug?  It seems like mypy should be able to handle normal usage of `lambda x: None`, so it might make sense to report the bug to upstream mypy and get it fixed there...

timabbott commented:
@cpdean just wanted to check in on this.  If you're busy, it'd be valuable to at least have an answer to the `nop_callback` question...

cpdean commented:
woops, hey I'll dig into the lambda thing tonight.  making commits better
will take me a bit - my git powers are pretty much just limited to squashing

On Mon, Jun 13, 2016, 12:42 Tim Abbott <notifications@github.com> wrote:

> @cpdean <https://github.com/cpdean> just wanted to check in on this. If
> you're busy, it'd be valuable to at least have an answer to the
> nop_callback question...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/958#issuecomment-225638178>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAJxYw62xU0gfdV94g58i7pN9vsVppiPks5qLYhngaJpZM4Iue-E>
> .
>

cpdean commented:
@timabbott okay, i've ripped out the nop_callback type things and I'm able to reproduce the error I saw earlier this month.

I boiled it down to a minimal example and reported the error here: https://github.com/python/mypy/issues/1710

Should we have the lambas in place and just add `# type: ignore   # https://github.com/python/mypy/issues/1710` for now?

cpdean commented:
woops, my bad. made a duplicate of https://github.com/python/mypy/issues/1425

timabbott commented:
No worries!  The `# type: ignore` approach linked to python/mypy#1425 sounds good.

cpdean commented:
added the ignore notes on inline lambdas and cleaned out a third of the commits

it'll be a couple days before i can take another pass but let me know what I should focus on. thanks @timabbott !

cpdean commented:
found some more stuff to clean up and make more clear. ready for feedback, thanks again @timabbott !

timabbott commented:
@sharmaeklavya2 can you do a review pass on this?

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -54,6 +56,7 @@ class ClientDescriptor(object):
     def __init__(self, user_profile_id, user_profile_email, realm_id, event_queue,
                  event_types, client_type_name, apply_markdown=True,
                  all_public_streams=False, lifespan_secs=0, narrow=[]):
+        # type: (int, str, int, EventQueue, List[str], str, Optional[bool], Optional[bool], Optional[int], Iterable[Sequence[text_type]]) -> None
```
`user_profile_email` should be `text_type`, isn't it?

sharmaeklavya2 replied:
Are you sure `None` is a valid value for `apply_markdown`, `all_public_streams` and `lifespan_secs`?

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -125,6 +133,7 @@ def add_event(self, event):
         self.finish_current_handler()
 
     def finish_current_handler(self, need_timeout=False):
+        # type: (Optional[bool]) -> bool
```
Shouldn't `need_timeout` be annotated as `bool`? I don't think `None` should be a valid value for `need_timeout`.

You might be mistaken about the meaning of `Optional`. `Optional[T]` is supposed to be the same as `Union[None, T]`.

cpdean replied:
ah, yeah i had assumed `Optional[T]` meant it was optional to the caller of the function, since the kwarg would set it to a default if not provided.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -146,21 +156,25 @@ def accepts_event(self, event):
                 and now - self.last_connection_time >= self.queue_timeout)
 
     def connect_handler(self, handler_id, client_name):
+        # type: (int, str) -> None
```
Is `client_name` of the same type as [zerver.models.Client.name](https://github.com/zulip/zulip/blob/2080038e7c5fb2fdcaeb06a9b14d29e01e6535e0/zerver/models.py#L643)? If yes, then it should be `text_type`. @timabbott maybe you know this.

sharmaeklavya2 replied:
I did some more grepping and it turns out that they are indeed the same. `client_name` should be `text_type`.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -170,6 +184,7 @@ def timeout_callback():
             self._timeout_handle = ioloop.add_timeout(heartbeat_time, timeout_callback)
 
     def disconnect_handler(self, client_closed=False, need_timeout=True):
+        # type: (Optional[bool], Optional[bool]) -> None
```
The parameters shouldn't be `Optional`.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -198,15 +214,19 @@ def cleanup(self):
     descriptors_by_handler_id[handler_id] = client_descriptor
 
 def clear_descriptor_by_handler_id(handler_id, client_descriptor):
+    # type: (int, ClientDescriptor) -> None
```
Here `client_descriptor` can be `None`. Therefore, `client_descriptor` should be `Optional[ClientDescriptor]`.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -343,7 +376,9 @@ def allocate_client_descriptor(new_queue_data):
     return client
 
 def do_gc_event_queues(to_remove, affected_users, affected_realms):
+    # type: (Set[str], Set[int], Set[int]) -> None
```
Replace `Set` by [`AbstractSet`](https://docs.python.org/3/library/typing.html#typing.AbstractSet) here. It's usually better to use abstract base types as type of parameters (see [cheat sheet](https://github.com/python/mypy/blob/master/docs/source/cheat_sheet.rst#standard-duck-types)).

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -343,7 +376,9 @@ def allocate_client_descriptor(new_queue_data):
 def do_gc_event_queues(to_remove, affected_users, affected_realms):
+    # type: (Set[str], Set[int], Set[int]) -> None
     def filter_client_dict(client_dict, key):
+        # type: (Dict, Union[int, str]) -> None
```
I think `client_dict` should be typed as `MutableMapping[Union[int, str], ClientDescriptor]`.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -94,10 +98,12 @@ def to_dict(self):
 
     @classmethod
     def from_dict(cls, d):
+        # type: (Dict[str, Any]) -> ClientDescriptor
```
I think `d` should be typed as `MutableMapping[str, Any]` instead. It is better to use duck types in parameters. See [cheat sheet](https://github.com/python/mypy/blob/master/docs/source/cheat_sheet.rst#standard-duck-types).

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -113,10 +119,12 @@ def from_dict(cls, d):
         self._timeout_handle = None
 
     def add_event(self, event):
+        # type: (Dict[str, Any]) -> None
```
`event` should be typed as `Mapping[str, Any]`.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -138,6 +147,7 @@ def finish_current_handler(self, need_timeout=False):
         return False
 
     def accepts_event(self, event):
+        # type: (Dict[str, Any]) -> bool
```
event should be `Mapping[str, Any]`.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -198,15 +214,19 @@ def cleanup(self):
     del descriptors_by_handler_id[handler_id]
 
 def compute_full_event_type(event):
+    # type: (Dict[str, Any]) -> str
```
`event` should be `Mapping[str, Any]`.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -233,13 +254,15 @@ def to_dict(self):
 
     @classmethod
     def from_dict(cls, d):
+        # type: (Dict[str, Any]) -> EventQueue
```
`d` should be `Mapping[str, Any]`.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -233,13 +254,15 @@ def to_dict(self):
         return ret
 
     def push(self, event):
+        # type: (Dict[str, Any]) -> None
```
event should be `MutableMapping[str, Any]`

sharmaeklavya2 replied:
Maybe that won't work because we are appending `event` to `self.queue` which is of type `deque[Dict[str, Any]]`. We can't append a `MutableSequence` to a deque of `Dict`s.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -424,6 +462,7 @@ def load_event_queues():
                  % (len(clients), time.time() - start))
 
 def send_restart_events(immediate=False):
+    # type: (Optional[bool]) -> None
```
immediate should be `bool`.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -453,6 +493,7 @@ def setup_event_queue():
     send_restart_events(immediate=settings.DEVELOPMENT)
 
 def fetch_events(query):
+    # type: (Dict[str, Any]) -> Dict[str, Any]
```
query should probably be `Mapping[str, Any]`.

cpdean replied:
Strange... when I switch this to Mapping I get the following error:

```
zerver/tornadoviews.py: note: In function "get_events_backend":
zerver/tornadoviews.py:96: error: Argument 1 to "fetch_events" has incompatible type Dict[str, Any]; expected Mapping[unicode, Any]
```

I would expect you could pass a Dict into an argument annotated as a Mapping, since Mapping is the more generic, duck-typed form of a Dict, right?

sharmaeklavya2 replied:
`Dict[str, Any]` has `str` as key type but `MutableMapping[unicode, Any]` has `unicode` as key type.

cpdean replied:
Ah, yeah when I switch the annotation to `MutableMapping[str, Any]` it passes.

sharmaeklavya2 replied:
I don't think `query` should be `Mapping` instead of `MutableMapping`. As far as I can see, we're not modifying `query` in this function.

cpdean replied:
Makes sense. Fixed.

sharmaeklavya2 replied:
I had accidentally written a wrong comment. I have fixed that comment.

cpdean replied:
@sharmaeklavya2 ah, yeah i changed it to `Mapping` since nothing was changing the `query`

cpdean commented:
@sharmaeklavya2 this is great thank you. admittedly I hadn't read the cheat sheet before starting this. I'll be able to clean this up this weekend.

sharmaeklavya2 commented:
Well I have suggested using `Mapping` or `MutableMapping` at some places, but there is a slight chance that might not work. For e.g., if an object has a property that is typed as a `Dict`, assigning a `Mapping` to it will fail. I have not checked your annotations very deeply so I might miss out on things like that.

After annotating all this code, you probably know it well enough that you can figure it out yourself what should be the correct type for a variable among `Mapping`, `MutableMapping` and `Dict`. But if you face any problems in doing so, just mention me and I'll be there to help.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -590,11 +639,12 @@ def missedmessage_hook(user_profile_id, queue, last_for_client):
+            queue_json_publish("missedmessage_emails", notice, nop)
 
 def receiver_is_idle(user_profile_id, realm_presences):
+    # type: (int, Dict[int, Dict]) -> bool
```
`realm_presences` should be `Optional[Mapping[int, Mapping]]`.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -629,6 +679,7 @@ def receiver_is_idle(user_profile_id, realm_presences):
     return off_zulip or idle
 
 def process_message_event(event_template, users):
```
event_template should be `Mapping[str, Any]`.
users should be `Iterable[Mapping[str, Any]]`.

cpdean replied:
What's the difference between `str` and `text_type`?  Should I just prefer `text_type` for everything?

sharmaeklavya2 replied:
`six.text_type` is `unicode` on python 2 and `str` on python 3. You should read more about this here - http://zulip.readthedocs.io/en/latest/mypy.html#annotating-strings.

sharmaeklavya2 commented on zerver/lib/queue.py:
```diff
@@ -11,14 +13,15 @@
+        # type: () -> None
         self.log = logging.getLogger('zulip.queue')
         self.queues = set() # type: Set[str]
         self.channel = None # type: Any
```
I think this should be typed as `Optional[BlockingChannel]`, since line 37 (in `_connect`) is
```py
        self.channel    = self.connection.channel() # type: BlockingChannel
```
and line 40 (in `_reconnect`) is
```py
self.channel = None
```

sharmaeklavya2 commented on zerver/lib/queue.py:
```diff
@@ -244,6 +276,7 @@ def wrapped_consumer(ch, method, properties, body):
 
 queue_client = None # type: SimpleQueueClient
```
`queue_client` should be an `Optional[SimpleQueueClient]`.

sharmaeklavya2 commented on zerver/lib/queue.py:
```diff
@@ -266,7 +300,7 @@ def setup_tornado_rabbitmq():
 
 def queue_json_publish(queue_name, event, processor):
-    # type: (str, Dict[str, Any], Callable) -> None
+    # type: (str, Dict[str, Any], Callable[[Any], None]) -> None
```
I think `event` should be `Mapping[str, Any]` and `processor` should be `Callable[[Mapping[str, Any]], None]`.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -463,9 +463,9 @@ def load_event_queues():
 def send_restart_events(immediate=False):
     # type: (Optional[bool]) -> None
-    event = dict(type='restart', server_generation=settings.SERVER_GENERATION)
+    event = dict(type='restart', server_generation=settings.SERVER_GENERATION)  # type: Dict[text_type, Any]
```
I think this should be `Dict[str, Any]`, since keyword argument names are `str`s.

sharmaeklavya2 commented on zerver/lib/queue.py:
```diff
@@ -3,7 +3,7 @@
 import pika
 from pika.adapters.blocking_connection import BlockingChannel
-from pika.spec import Deliver, BasicProperties
+from pika.spec import Basic, BasicProperties
```
I think you have consistently used `pika.BasicProperties` in your code. So you can probably change it to `from pika.spec import Basic`.

sharmaeklavya2 commented:
@timabbott @cpdean I have reviewed this pull request.

cpdean commented:
@sharmaeklavya2 i had missed your comment about `pika.BasicProperties` earlier but have fixed it. Also I fixed the annotation on `event_types` to set it to `List[str]` since the other parts of Zulip seem to agree on that type.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -56,7 +56,7 @@ class ClientDescriptor(object):
                  event_types, client_type_name, apply_markdown=True,
                  all_public_streams=False, lifespan_secs=0, narrow=[]):
-        # type: (int, text_type, int, EventQueue, List[text_type], text_type, bool, bool, int, Iterable[Sequence[text_type]]) -> None
+        # type: (int, text_type, int, EventQueue, List[str], text_type, bool, bool, int, Iterable[Sequence[text_type]]) -> None
```
I see that `self.event_types` is never modified once it is set (not just in this function, but also outside it), so maybe `Sequence[str]` would be better here.
Actually there are some places where `self.event_types` is being compared to `None`, like in `ClientDescriptor.accepts_event`. So it should be `Optional[Sequence[str]]`.

cpdean replied:
Oh good catch. Fixed it.

sharmaeklavya2 commented on zerver/lib/event_queue.py:
```diff
@@ -65,7 +65,7 @@ def __init__(self, user_profile_id, user_profile_email, realm_id, event_queue,
         self.realm_id = realm_id
         self.current_handler_id = None # type: int
-        self.current_client_name = None # type: str
+        self.current_client_name = None # type: text_type
```
`self.current_client_name` should be `Optional[text_type]`.

cpdean replied:
Gotcha. pushed up now.

sharmaeklavya2 commented on zerver/tests/test_subs.py:
```diff
@@ -781,7 +781,7 @@ def test_successful_subscriptions_add(self):
         """
         self.assertNotEqual(len(self.streams), 0)  # necessary for full test coverage
-        add_streams = [u"Verona2", u"Denmark5"]
+        add_streams = [u"Verona2", u"Denmark5"] # type: List[text_type]
```
Is this really needed? Isn't mypy able to deduce this? (I haven't checked)

sharmaeklavya2 commented on zerver/decorator.py:
```diff
@@ -70,7 +70,7 @@ def update_user_activity(request, user_profile):
            'time': datetime_to_timestamp(now()),
            'client': request.client.name}
-    queue_json_publish("user_activity", event, lambda event: None)
+    queue_json_publish("user_activity", event, lambda x: None) # type: ignore  # waiting on https://github.com/python/mypy/issues/1425
```
This issue has been resolved and we have upgraded zulip's mypy version to be new enough to not suffer from this.

timabbott commented:
Merged via #1171.  Thanks @cpdean for working on this!!
